source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,166 @@ namespace aspect
+                || max_advection_local > max_advection_exact_global)
+              {
+                //Compute the cell area and cell advection average
+                double local_area = 0;
+                double local_advection_average = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_area += fe_values_12.JxW(q);
+                    local_advection_average += values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_advection_average /= local_area;
+                /*
+                 * Define theta: a scaling constant used to correct the old solution by the formula
+                 *   new_value = theta * (old_value-old_solution_cell_average)+old_solution_cell_average
+                 * where theta \in [0,1] defined as below.
+                 * After the correction, the new solution does not exceed the user-given
+                 * exact global maximum/minimum values. Meanwhile, the new solution's cell average
+                 * equals to the old solution's cell average.
+                 */
+                double theta = std::min<double>
+                               (1, abs((max_advection_exact_global-local_advection_average)
+                                       /(max_advection_local-local_advection_average)));
+                theta = std::min<double>
+                        (theta, abs((min_advection_exact_global-local_advection_average)
+                                    /(min_advection_local-local_advection_average)));
+                //Modify the numerical solution at freedoms of advection field
+                for (unsigned int j = 0;
+                     j < finite_element.base_element(advection_field.base_element(introspection)).dofs_per_cell;
+                     ++j)
+                  {
+                    const unsigned int support_point_index = finite_element.component_to_system_index(
+                                                               (advection_field.is_temperature()
+                                                                ?
+                                                                introspection.component_indices.temperature
+                                                                :
+                                                                introspection.component_indices.compositional_fields[advection_field.compositional_variable]
+                                                               ),
+                                                               /*dof index within component=*/ j);
+                    const double solution_value = distributed_vector(local_dof_indices[support_point_index]);
+                    const double limited_solution_value = theta * (solution_value-local_advection_average) + local_advection_average;
+                    distributed_vector(local_dof_indices[support_point_index]) = limited_solution_value;
```

Comment:
have you run this in parallel? I suspect that you can actually only do this if `local_dof_indices[support_point_index]` is a locally owned dof (not all dofs on locally owned cells are in fact locally owned dofs).

Or, on second thought, you probably can do the write but don't need to. The `compress` call below does the data exchange. What you probably can't do is the read in line 1060. I think it would be useful to only do the last three lines if `support_point_index` is a locally owned dof. There is an `IndexSet` variable somewhere in the `Simulator` object that you can query whether that is the case.